base/compiler/optimize.jl

```diff
@@ -211,88 +214,81 @@ Returns a tuple of (effect_free_and_nothrow, nothrow) for a given statement.
             # if we aren't certain enough about the type, it might be an UndefVarError at runtime
             nothrow = isa(etyp, Const)
-            return (nothrow, nothrow)
+            return (true, nothrow, nothrow)
         end
         if head === :call
             f = argextype(args[1], src)
             f = singleton_type(f)
-            f === nothing && return (false, false)
-            if isa(f, IntrinsicFunction)
-                nothrow = intrinsic_nothrow(f,
-                    Any[argextype(args[i], src) for i = 2:length(args)])
-                nothrow || return (false, false)
-                return (intrinsic_effect_free_if_nothrow(f), nothrow)
-            end
-            contains_is(_PURE_BUILTINS, f) && return (true, true)
-            # `get_binding_type` sets the type to Any if the binding doesn't exist yet
-            if f === Core.get_binding_type
-                length(args) == 3 || return false
-                M, s = argextype(args[2], src), argextype(args[3], src)
-                total = get_binding_type_effect_free(M, s)
-                return (total, total)
-            end
-            rt === Bottom && return (false, false)
-            nothrow = _builtin_nothrow(f, Any[argextype(args[i], src) for i = 2:length(args)], rt, OptimizerLattice())
-            nothrow || return (false, false)
-            return (contains_is(_EFFECT_FREE_BUILTINS, f), nothrow)
+            f === nothing && return (false, false, false)
+            isa(f, Builtin) || return (false, false, false)
+            # Needs to be handled in inlining to look at the callee effects
+            f === Core._apply_iterate && return (false, false, false)
+            argtypes = Any[argextype(args[arg], src) for arg in 2:length(args)]
+            effects = builtin_effects(f, argtypes, rt)
+            consistent = is_consistent(effects)
+            effect_free = is_effect_free(effects)
+            nothrow = is_nothrow(effects)
+            return (consistent, effect_free & nothrow, nothrow)
         elseif head === :new
             typ = argextype(args[1], src)
             # `Expr(:new)` of unknown type could raise arbitrary TypeError.
             typ, isexact = instanceof_tfunc(typ)
-            isexact || return (false, false)
-            isconcretedispatch(typ) || return (false, false)
+            isexact || return (false, false, false)
+            isconcretedispatch(typ) || return (false, false, false)
             typ = typ::DataType
-            fieldcount(typ) >= length(args) - 1 || return (false, false)
+            fieldcount(typ) >= length(args) - 1 || return (false, false, false)
             for fld_idx in 1:(length(args) - 1)
                 eT = argextype(args[fld_idx + 1], src)
                 fT = fieldtype(typ, fld_idx)
-                eT ⊑ₒ fT || return (false, false)
+                eT ⊑ₒ fT || return (false, false, false)
             end
-            return (true, true)
+            return (true, true, true)
```

Comment:
I think we should return `(false, true, true)` for mutable allocation?

Reply:
That is certainly the more conservative option - will update.